coderd/activitybump.go

```diff
@@ -0,0 +1,77 @@
+package coderd
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/coderd/database"
+)
+
+// activityBumpWorkspace automatically bumps the workspace's auto-off timer
+// if it is set to expire soon.
+func activityBumpWorkspace(log slog.Logger, db database.Store, workspace database.Workspace) {
+	// We set a short timeout so if the app is under load, these
+	// low priority operations fail first.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	err := db.InTx(func(s database.Store) error {
+		build, err := s.GetLatestWorkspaceBuildByWorkspaceID(ctx, workspace.ID)
+		log.Debug(ctx, "build", slog.F("build", build))
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil
+		} else if err != nil {
+			return xerrors.Errorf("get latest workspace build: %w", err)
+		}
+
+		job, err := s.GetProvisionerJobByID(ctx, build.JobID)
+		if err != nil {
+			return xerrors.Errorf("get provisioner job: %w", err)
+		}
+
+		if build.Transition != database.WorkspaceTransitionStart || !job.CompletedAt.Valid {
+			return nil
+		}
+
+		if build.Deadline.IsZero() {
+			// Workspace shutdown is manual
+			return nil
+		}
+
+		// We sent bumpThreshold slightly under bumpAmount to minimize DB writes.
+		const (
+			bumpAmount    = time.Hour
+			bumpThreshold = time.Hour - (time.Minute * 10)
+		)
+
+		if !build.Deadline.Before(time.Now().Add(bumpThreshold)) {
+			return nil
+		}
+
+		newDeadline := time.Now().Add(bumpAmount)
+
+		if err := s.UpdateWorkspaceBuildByID(ctx, database.UpdateWorkspaceBuildByIDParams{
+			ID:               build.ID,
+			UpdatedAt:        build.UpdatedAt,
```

Comment:
We should update this since we're modifying data.
```suggestion
			UpdatedAt:        database.Now(),
```